File: experiment-5/runs/language=typescript_model=claude-opus-4-7_tooling=beads/rep2/src/queries/competitions.ts
import type { Match, TeamRecord } from "../types.js";

export interface StandingsFilter {
  competition: string;
  season: number;
}

export function standings(matches: Match[], filter: StandingsFilter): TeamRecord[] {
  const compNeedle = filter.competition.toLowerCase();
  const tally = new Map<string, TeamRecord & { displayName: string }>();

  function getRow(displayName: string): TeamRecord & { displayName: string } {
    const key = displayName.toLowerCase();
    let r = tally.get(key);
    if (!r) {
      r = {
        team: displayName,
        displayName,
        matches: 0,
        wins: 0,
        draws: 0,
        losses: 0,
        goalsFor: 0,
        goalsAgainst: 0,
        goalDifference: 0,
        points: 0,
        winRate: 0,
      };
      tally.set(key, r);
    }
    return r;
  }

  for (const m of matches) {
    if (m.season !== filter.season) continue;
    if (!String(m.competition).toLowerCase().includes(compNeedle)) continue;
    if (m.homeGoal === null || m.awayGoal === null) continue;

    const home = getRow(m.homeTeam);
    const away = getRow(m.awayTeam);

    home.matches++; away.matches++;
    home.goalsFor += m.homeGoal; home.goalsAgainst += m.awayGoal;
    away.goalsFor += m.awayGoal; away.goalsAgainst += m.homeGoal;

    if (m.homeGoal > m.awayGoal) { home.wins++; away.losses++; }
    else if (m.homeGoal < m.awayGoal) { away.wins++; home.losses++; }
    else { home.draws++; away.draws++; }
  }

  const rows = Array.from(tally.values()).map((r) => {
    r.points = r.wins * 3 + r.draws;
    r.goalDifference = r.goalsFor - r.goalsAgainst;
    r.winRate = r.matches > 0 ? r.wins / r.matches : 0;
    return r;
  });

  rows.sort((a, b) => {
    if (b.points !== a.points) return b.points - a.points;
    if (b.wins !== a.wins) return b.wins - a.wins;
    if (b.goalDifference !== a.goalDifference) return b.goalDifference - a.goalDifference;
    if (b.goalsFor !== a.goalsFor) return b.goalsFor - a.goalsFor;
    return a.team.localeCompare(b.team);
  });

  return rows;
}

export function champion(matches: Match[], filter: StandingsFilter): TeamRecord | null {
  const s = standings(matches, filter);
  return s.length > 0 ? s[0] : null;
}

export function relegated(matches: Match[], filter: StandingsFilter, count = 4): TeamRecord[] {
  const s = standings(matches, filter);
  return s.slice(Math.max(0, s.length - count));
}

export function listCompetitions(matches: Match[]): string[] {
  const set = new Set<string>();
  for (const m of matches) {
    set.add(String(m.competition));
  }
  return Array.from(set).sort();
}

export function listSeasons(matches: Match[], competition?: string): number[] {
  const needle = competition?.toLowerCase();
  const set = new Set<number>();
  for (const m of matches) {
    if (needle && !String(m.competition).toLowerCase().includes(needle)) continue;
    if (m.season !== null) set.add(m.season);
  }
  return Array.from(set).sort((a, b) => a - b);
}

File: experiment-14/runs/language=elixir_model=opus-4.8-fast_prompt=TDD/rep1-failed/lib/brasil_soccer/competitions.ex
defmodule BrasilSoccer.Competitions do
  @moduledoc """
  Competition-level views derived from match results: league standings (points,
  goal difference, ordering), the implied champion, and the seasons available.

  Standings use the conventional three-points-for-a-win system and break ties on
  goal difference, then goals scored.
  """

  alias BrasilSoccer.{Match, Matches}

  @doc """
  Compute the standings table for `competition` in `season`. Returns a list of
  team rows ordered best-first, each carrying a `:position`.
  """
  @spec standings([Match.t()], String.t(), integer()) :: [map()]
  def standings(matches, competition, season) do
    matches
    |> Matches.find(competition: competition, season: season)
    |> Enum.reduce(%{}, &accumulate/2)
    |> Map.values()
    |> Enum.map(&finalize/1)
    |> Enum.sort_by(&{&1.points, &1.goal_difference, &1.goals_for}, :desc)
    |> Enum.with_index(1)
    |> Enum.map(fn {row, pos} -> Map.put(row, :position, pos) end)
  end

  @doc "The champion (top of the table), or `nil` if there are no matches."
  @spec champion([Match.t()], String.t(), integer()) :: map() | nil
  def champion(matches, competition, season) do
    case standings(matches, competition, season) do
      [first | _] -> first
      [] -> nil
    end
  end

  @doc "Distinct seasons present for a competition, ascending."
  @spec seasons([Match.t()], String.t()) :: [integer()]
  def seasons(matches, competition) do
    matches
    |> Matches.find(competition: competition)
    |> Enum.map(& &1.season)
    |> Enum.reject(&is_nil/1)
    |> Enum.uniq()
    |> Enum.sort()
  end

  # Only count matches with a known scoreline towards the table.
  defp accumulate(%Match{winner: nil}, acc), do: acc

  defp accumulate(%Match{} = match, acc) do
    acc
    |> add_side(match, :home)
    |> add_side(match, :away)
  end

  defp add_side(acc, match, side) do
    {team, gf, ga} = side_fields(match, side)

    row =
      Map.get(acc, team, %{
        team: team,
        played: 0,
        wins: 0,
        draws: 0,
        losses: 0,
        goals_for: 0,
        goals_against: 0,
        points: 0
      })

    row =
      row
      |> Map.update!(:played, &(&1 + 1))
      |> Map.update!(:goals_for, &(&1 + gf))
      |> Map.update!(:goals_against, &(&1 + ga))
      |> apply_result(result(match, side))

    Map.put(acc, team, row)
  end

  defp side_fields(match, :home), do: {match.home_team, match.home_goal, match.away_goal}
  defp side_fields(match, :away), do: {match.away_team, match.away_goal, match.home_goal}

  defp result(%Match{winner: :draw}, _side), do: :draw
  defp result(%Match{winner: side}, side), do: :win
  defp result(_match, _side), do: :loss

  defp apply_result(row, :win),
    do: row |> Map.update!(:wins, &(&1 + 1)) |> Map.update!(:points, &(&1 + 3))

  defp apply_result(row, :draw),
    do: row |> Map.update!(:draws, &(&1 + 1)) |> Map.update!(:points, &(&1 + 1))

  defp apply_result(row, :loss), do: Map.update!(row, :losses, &(&1 + 1))

  defp finalize(row), do: Map.put(row, :goal_difference, row.goals_for - row.goals_against)
end
